readme.md
**Adding Glasses and Mustache to Rotated Faces with Mediapipe**

This Python program uses the Mediapipe library to detect facial landmarks and add glasses and mustache to the face, even when it's rotated left or right. The program works by:

Capturing a video stream from the default camera or a video file.
Using the Mediapipe FaceMesh solution to detect 468 facial landmarks on each frame of the video.
Calculating the angle of rotation of the face based on the position of the landmarks around the eyes and nose.
Applying a rotation transformation to the glasses and mustache images to match the angle of the face.
Overlaying the glasses and mustache images on top of the face using the landmarks around the eyes and mouth as reference points.
Displaying the resulting video stream with the added glasses and mustache.
Requirements
To run this program, you need to have the following installed:
<table>
  <tr>
    <td><img src="https://user-images.githubusercontent.com/114035408/235405259-18edf1ed-ffe4-4fe7-a3c3-628014028868.jpg" alt="Image 1" width="200" height="200"></td>
    <td><img src="https://user-images.githubusercontent.com/114035408/235405264-2224fc4f-4819-4d71-b1a7-c4402175337a.jpg" alt="Image 2" width="200" height="200"></td>
    <td><img src="https://user-images.githubusercontent.com/114035408/235405270-6a914251-20c8-4626-868d-01c6f07e0e57.jpg" alt="Image 3" width="200" height="200"></td>
    <td><img src="https://user-images.githubusercontent.com/114035408/235405273-a365a7b1-bdba-422d-9590-1b93f18375bd.jpg" alt="Image 3" width="200" height="200"></td>
  </tr>
</table>

Python 3.6 or later
OpenCV 4.5 or later
Mediapipe 0.8 or later


Usage:
To run this program, open a teminal and run apply_filter.py


history.md
Contribution: 2019-07-02 20:00

Contribution: 2019-07-02 20:01

Contribution: 2019-07-02 20:02

Contribution: 2019-07-02 20:03

Contribution: 2019-07-02 20:04

Contribution: 2019-07-02 20:05

Contribution: 2019-07-02 20:06

Contribution: 2019-07-02 20:07

Contribution: 2019-07-02 20:08

Contribution: 2019-07-02 20:09

Contribution: 2019-07-04 20:00

Contribution: 2019-07-04 20:01

Contribution: 2019-07-04 20:02

Contribution: 2019-07-04 20:03

Contribution: 2019-07-04 20:04

Contribution: 2019-07-09 20:00

Contribution: 2019-07-09 20:01

Contribution: 2019-07-09 20:02

Contribution: 2019-07-18 20:00

Contribution: 2019-07-18 20:01

Contribution: 2019-07-18 20:02

Contribution: 2019-07-18 20:03

Contribution: 2019-07-18 20:04

Contribution: 2019-07-18 20:05

Contribution: 2019-07-18 20:06

Contribution: 2019-07-18 20:07

Contribution: 2019-07-19 20:00

Contribution: 2019-07-19 20:01

Contribution: 2019-07-19 20:02

Contribution: 2019-07-19 20:03

Contribution: 2019-07-21 20:00

Contribution: 2019-07-21 20:01

Contribution: 2019-07-21 20:02

Contribution: 2019-07-27 20:00

Contribution: 2019-07-27 20:01

Contribution: 2019-07-27 20:02

Contribution: 2019-07-27 20:03

Contribution: 2019-07-27 20:04

Contribution: 2019-07-27 20:05

Contribution: 2019-07-27 20:06

Contribution: 2019-07-27 20:07

Contribution: 2019-07-31 20:00

Contribution: 2019-07-31 20:01

Contribution: 2019-07-31 20:02

Contribution: 2019-07-31 20:03

Contribution: 2019-07-31 20:04

Contribution: 2019-07-31 20:05

Contribution: 2019-07-31 20:06

Contribution: 2019-08-01 20:00

Contribution: 2019-08-02 20:00

Contribution: 2019-08-02 20:01

Contribution: 2019-08-02 20:02

Contribution: 2019-08-02 20:03

Contribution: 2019-08-02 20:04

Contribution: 2019-08-02 20:05

Contribution: 2019-08-02 20:06

Contribution: 2019-08-02 20:07

Contribution: 2019-08-02 20:08

Contribution: 2019-08-02 20:09

Contribution: 2019-08-04 20:00

Contribution: 2019-08-04 20:01

Contribution: 2019-08-04 20:02

Contribution: 2019-08-04 20:03

Contribution: 2019-08-04 20:04

Contribution: 2019-08-04 20:05

Contribution: 2019-08-07 20:00

Contribution: 2019-08-09 20:00

Contribution: 2019-08-09 20:01

Contribution: 2019-08-12 20:00

Contribution: 2019-08-12 20:01

Contribution: 2019-08-12 20:02

Contribution: 2019-08-12 20:03

Contribution: 2019-08-12 20:04

Contribution: 2019-08-12 20:05

Contribution: 2019-08-12 20:06

Contribution: 2019-08-12 20:07

Contribution: 2019-08-12 20:08

Contribution: 2019-08-19 20:00

Contribution: 2019-08-19 20:01

Contribution: 2019-08-19 20:02

Contribution: 2019-08-20 20:00

Contribution: 2019-08-20 20:01

Contribution: 2019-08-20 20:02

Contribution: 2019-08-20 20:03

Contribution: 2019-08-21 20:00

Contribution: 2019-08-22 20:00

Contribution: 2019-08-22 20:01

Contribution: 2019-08-22 20:02

Contribution: 2019-08-22 20:03

Contribution: 2019-08-22 20:04

Contribution: 2019-08-22 20:05

Contribution: 2019-08-23 20:00

Contribution: 2019-08-23 20:01

Contribution: 2019-08-23 20:02

Contribution: 2019-08-23 20:03

Contribution: 2019-08-23 20:04

Contribution: 2019-08-23 20:05

Contribution: 2019-08-23 20:06

Contribution: 2019-08-23 20:07

Contribution: 2019-08-24 20:00

Contribution: 2019-08-24 20:01

Contribution: 2019-08-24 20:02

Contribution: 2019-08-24 20:03

Contribution: 2019-08-24 20:04

Contribution: 2019-08-28 20:00

Contribution: 2019-08-28 20:01

Contribution: 2019-08-28 20:02

Contribution: 2019-08-28 20:03

Contribution: 2019-08-28 20:04

Contribution: 2019-08-28 20:05

Contribution: 2019-08-29 20:00

Contribution: 2019-08-29 20:01

Contribution: 2019-08-29 20:02

Contribution: 2019-08-29 20:03

Contribution: 2019-08-29 20:04

Contribution: 2019-08-29 20:05

Contribution: 2019-08-29 20:06

Contribution: 2019-08-29 20:07

Contribution: 2019-08-30 20:00

Contribution: 2019-08-30 20:01

Contribution: 2019-08-30 20:02

Contribution: 2019-08-31 20:00

Contribution: 2019-09-02 20:00

Contribution: 2019-09-02 20:01

Contribution: 2019-09-02 20:02

Contribution: 2019-09-02 20:03

Contribution: 2019-09-02 20:04

Contribution: 2019-09-02 20:05

Contribution: 2019-09-06 20:00

Contribution: 2019-09-06 20:01

Contribution: 2019-09-06 20:02

Contribution: 2019-09-06 20:03

Contribution: 2019-09-06 20:04

Contribution: 2019-09-06 20:05

Contribution: 2019-09-06 20:06

Contribution: 2019-09-06 20:07

Contribution: 2019-09-06 20:08

Contribution: 2019-09-06 20:09

Contribution: 2019-09-07 20:00

Contribution: 2019-09-07 20:01

Contribution: 2019-09-07 20:02

Contribution: 2019-09-07 20:03

Contribution: 2019-09-07 20:04

Contribution: 2019-09-07 20:05

Contribution: 2019-09-09 20:00

Contribution: 2019-09-09 20:01

Contribution: 2019-09-09 20:02

Contribution: 2019-09-09 20:03

Contribution: 2019-09-09 20:04

Contribution: 2019-09-09 20:05

Contribution: 2019-09-09 20:06

Contribution: 2019-09-11 20:00

Contribution: 2019-09-11 20:01

Contribution: 2019-09-11 20:02

Contribution: 2019-09-11 20:03

Contribution: 2019-09-11 20:04

Contribution: 2019-09-11 20:05

Contribution: 2019-09-11 20:06

Contribution: 2019-09-12 20:00

Contribution: 2019-09-12 20:01

Contribution: 2019-09-12 20:02

Contribution: 2019-09-12 20:03

Contribution: 2019-09-12 20:04

Contribution: 2019-09-12 20:05

Contribution: 2019-09-12 20:06

Contribution: 2019-09-12 20:07

Contribution: 2019-09-14 20:00

Contribution: 2019-09-14 20:01

Contribution: 2019-09-14 20:02

Contribution: 2019-09-14 20:03

Contribution: 2019-09-14 20:04

Contribution: 2019-09-14 20:05

Contribution: 2019-09-19 20:00

Contribution: 2019-09-20 20:00

Contribution: 2019-09-20 20:01

Contribution: 2019-09-20 20:02

Contribution: 2019-09-20 20:03

Contribution: 2019-09-20 20:04

Contribution: 2019-09-20 20:05

Contribution: 2019-09-20 20:06

Contribution: 2019-09-30 20:00

Contribution: 2019-09-30 20:01

Contribution: 2019-09-30 20:02

Contribution: 2019-09-30 20:03

Contribution: 2019-09-30 20:04

Contribution: 2019-09-30 20:05

Contribution: 2019-09-30 20:06


facemesh_test.py
import cv2
import mediapipe as mp

mp_drawing = mp.solutions.drawing_utils
mp_drawing_styles = mp.solutions.drawing_styles
mp_face_mesh = mp.solutions.face_mesh

keypnt_68 = [127, 93, 58, 136, 150, 149, 176, 148, 152, 377, 400, 378, 379, 365, 288, 323, 356, 70, 63, 105, 66, 55,
             285, 296, 334, 293, 300, 168, 6, 195, 4, 64, 60, 94, 290, 439, 33, 160, 158, 173, 153, 144, 398, 385, 387,
             466, 373, 380, 61, 40, 39, 0, 269, 270, 291, 321, 405, 17, 181, 91, 78, 81, 13, 311, 306, 402, 14, 178,
             162, 54, 67, 10, 297, 284, 389]

# For static images:
IMAGE_FILES = []
drawing_spec = mp_drawing.DrawingSpec(thickness=1, circle_radius=1)
with mp_face_mesh.FaceMesh(
        static_image_mode=True,
        max_num_faces=1,
        refine_landmarks=True,
        min_detection_confidence=0.5) as face_mesh:
    for idx, file in enumerate(IMAGE_FILES):
        image = cv2.imread(file)
        # Convert the BGR image to RGB before processing.
        results = face_mesh.process(cv2.cvtColor(image, cv2.COLOR_BGR2RGB))

        # Print and draw face mesh landmarks on the image.
        if not results.multi_face_landmarks:
            continue
        annotated_image = image.copy()
        for face_landmarks in results.multi_face_landmarks:
            print('face_landmarks:', face_landmarks)
            mp_drawing.draw_landmarks(
                image=annotated_image,
                landmark_list=face_landmarks,
                connections=mp_face_mesh.FACEMESH_TESSELATION,
                landmark_drawing_spec=None,
                connection_drawing_spec=mp_drawing_styles
                    .get_default_face_mesh_tesselation_style())
            mp_drawing.draw_landmarks(
                image=annotated_image,
                landmark_list=face_landmarks,
                connections=mp_face_mesh.FACEMESH_CONTOURS,
                landmark_drawing_spec=None,
                connection_drawing_spec=mp_drawing_styles
                    .get_default_face_mesh_contours_style())
            mp_drawing.draw_landmarks(
                image=annotated_image,
                landmark_list=face_landmarks,
                connections=mp_face_mesh.FACEMESH_IRISES,
                landmark_drawing_spec=None,
                connection_drawing_spec=mp_drawing_styles
                    .get_default_face_mesh_iris_connections_style())
        cv2.imwrite('/tmp/annotated_image' + str(idx) + '.png', annotated_image)

# For webcam input:
drawing_spec = mp_drawing.DrawingSpec(thickness=1, circle_radius=1)
cap = cv2.VideoCapture(0)
with mp_face_mesh.FaceMesh(
        max_num_faces=1,
        refine_landmarks=True,
        min_detection_confidence=0.5,
        min_tracking_confidence=0.5) as face_mesh:
    while cap.isOpened():
        success, image = cap.read()
        if not success:
            print("Ignoring empty camera frame.")
            # If loading a video, use 'break' instead of 'continue'.
            continue

        # To improve performance, optionally mark the image as not writeable to
        # pass by reference.
        image.flags.writeable = False
        image = cv2.cvtColor(image, cv2.COLOR_BGR2RGB)
        results = face_mesh.process(image)

        # Draw the face mesh annotations on the image.
        image.flags.writeable = True
        image = cv2.cvtColor(image, cv2.COLOR_RGB2BGR)
        # image1 = image.copy()
        # image2 = image.copy()
        if results.multi_face_landmarks:
            contour_point_idxs = []
            for point in mp_face_mesh.FACEMESH_CONTOURS:
                contour_point_idxs.append(point[0])
            for face_landmarks in results.multi_face_landmarks:
                for idx, landmark in enumerate(face_landmarks.landmark):
                    # if idx in contour_point_idxs:
                    if idx in keypnt_68:
                        # if True:
                        height, width = image.shape[:2]
                        point = (int(landmark.x * width), int(landmark.y * height))
                        cv2.circle(image, point, 1, (255, 0, 0), -1)
                        cv2.putText(image, str(idx), point, cv2.FONT_HERSHEY_SIMPLEX, .3, (255, 255, 255), 1)
                        # if idx % 2 == 0:
                        #     cv2.circle(image2, point, 1, (255, 0, 0), -1)
                        #     cv2.putText(image2, str(idx), point, cv2.FONT_HERSHEY_SIMPLEX, .3, (255, 255, 255), 1)
                        # elif (idx+1) % 2 == 0:
                        #     cv2.circle(image1, point, 1, (255, 0, 0), -1)
                        #     cv2.putText(image1, str(idx), point, cv2.FONT_HERSHEY_SIMPLEX, .3, (255, 255, 255), 1)

        # Flip the image horizontally for a selfie-view display.
        # cv2.imshow('MediaPipe Face Mesh', cv2.flip(image, 1))
        cv2.imshow('MediaPipe Face Mesh', image)
        # cv2.imshow('image2', image2)
        # cv2.imshow('image1', image1)
        if cv2.waitKey(5) & 0xFF == 27:
            break
cap.release()

faceMesh.py
import cv2
import mediapipe as mp
mp_drawing = mp.solutions.drawing_utils
mp_drawing_styles = mp.solutions.drawing_styles
mp_face_mesh = mp.solutions.face_mesh

# For static images:
IMAGE_FILES = []
drawing_spec = mp_drawing.DrawingSpec(thickness=1, circle_radius=1)
with mp_face_mesh.FaceMesh(
    static_image_mode=True,
    max_num_faces=1,
    refine_landmarks=True,
    min_detection_confidence=0.5) as face_mesh:
  for idx, file in enumerate(IMAGE_FILES):
    image = cv2.imread(file)
    # Convert the BGR image to RGB before processing.
    results = face_mesh.process(cv2.cvtColor(image, cv2.COLOR_BGR2RGB))

    # Print and draw face mesh landmarks on the image.
    if not results.multi_face_landmarks:
      continue
    annotated_image = image.copy()
    for face_landmarks in results.multi_face_landmarks:
      print('face_landmarks:', face_landmarks)
      mp_drawing.draw_landmarks(
          image=annotated_image,
          landmark_list=face_landmarks,
          connections=mp_face_mesh.FACEMESH_TESSELATION,
          landmark_drawing_spec=None,
          connection_drawing_spec=mp_drawing_styles
          .get_default_face_mesh_tesselation_style())
      mp_drawing.draw_landmarks(
          image=annotated_image,
          landmark_list=face_landmarks,
          connections=mp_face_mesh.FACEMESH_CONTOURS,
          landmark_drawing_spec=None,
          connection_drawing_spec=mp_drawing_styles
          .get_default_face_mesh_contours_style())
      mp_drawing.draw_landmarks(
          image=annotated_image,
          landmark_list=face_landmarks,
          connections=mp_face_mesh.FACEMESH_IRISES,
          landmark_drawing_spec=None,
          connection_drawing_spec=mp_drawing_styles
          .get_default_face_mesh_iris_connections_style())
    cv2.imwrite('pipetest' + str(idx) + '.png', annotated_image)

# For webcam input:
drawing_spec = mp_drawing.DrawingSpec(thickness=1, circle_radius=1)
# cap = cv2.VideoCapture(0)
with mp_face_mesh.FaceMesh(
    max_num_faces=1,
    refine_landmarks=True,
    min_detection_confidence=0.5,
    min_tracking_confidence=0.5) as face_mesh:
  

  image = cv2.imread("custom/3.jpg")  
  # To improve performance, optionally mark the image as not writeable to
  # pass by reference.
  image.flags.writeable = False
  image = cv2.cvtColor(image, cv2.COLOR_BGR2RGB)
  results = face_mesh.process(image)

  # Draw the face mesh annotations on the image.
  image.flags.writeable = True
  image = cv2.cvtColor(image, cv2.COLOR_RGB2BGR)
  if results.multi_face_landmarks:
    for face_landmarks in results.multi_face_landmarks:
      # print(face_landmarks)
      mp_drawing.draw_landmarks(
          image=image,
          landmark_list=face_landmarks,
          connections=mp_face_mesh.FACEMESH_TESSELATION,
          landmark_drawing_spec=None,
          connection_drawing_spec=mp_drawing_styles
          .get_default_face_mesh_tesselation_style())
      mp_drawing.draw_landmarks(
          image=image,
          landmark_list=face_landmarks,
          connections=mp_face_mesh.FACEMESH_CONTOURS,
          landmark_drawing_spec=None,
          connection_drawing_spec=mp_drawing_styles
          .get_default_face_mesh_contours_style())
      mp_drawing.draw_landmarks(
          image=image,
          landmark_list=face_landmarks,
          connections=mp_face_mesh.FACEMESH_IRISES,
          landmark_drawing_spec=None,
          connection_drawing_spec=mp_drawing_styles
          .get_default_face_mesh_iris_connections_style())
  # Flip the image horizontally for a selfie-view display.
  cv2.imshow('MediaPipe Face Mesh', cv2.flip(image, 1))
  cv2.imwrite("result.jpg", image)
  cv2.waitKey()

faceBlendCommon.py
import cv2
import numpy as np
import math


# Constrains points to be inside boundary
def constrainPoint(p, w, h):
  p = (min(max(p[0], 0), w - 1), min(max(p[1], 0), h - 1))
  return p

# Compute similarity transform given two sets of two points.
# OpenCV requires 3 pairs of corresponding points.
# We are faking the third one.
def similarityTransform(inPoints, outPoints):
  s60 = math.sin(60*math.pi/180)
  c60 = math.cos(60*math.pi/180)

  inPts = np.copy(inPoints).tolist()
  outPts = np.copy(outPoints).tolist()

  # The third point is calculated so that the three points make an equilateral triangle
  xin = c60*(inPts[0][0] - inPts[1][0]) - s60*(inPts[0][1] - inPts[1][1]) + inPts[1][0]
  yin = s60*(inPts[0][0] - inPts[1][0]) + c60*(inPts[0][1] - inPts[1][1]) + inPts[1][1]

  inPts.append([np.int_(xin), np.int_(yin)])

  xout = c60*(outPts[0][0] - outPts[1][0]) - s60*(outPts[0][1] - outPts[1][1]) + outPts[1][0]
  yout = s60*(outPts[0][0] - outPts[1][0]) + c60*(outPts[0][1] - outPts[1][1]) + outPts[1][1]

  outPts.append([np.int_(xout), np.int_(yout)])

  # Now we can use estimateRigidTransform for calculating the similarity transform.
  tform = cv2.estimateAffinePartial2D(np.array([inPts]), np.array([outPts]))
  return tform[0]


# Check if a point is inside a rectangle
def rectContains(rect, point):
  if point[0] < rect[0]:
    return False
  elif point[1] < rect[1]:
    return False
  elif point[0] > rect[2]:
    return False
  elif point[1] > rect[3]:
    return False
  return True


# Calculate Delaunay triangles for set of points
# Returns the vector of indices of 3 points for each triangle
def calculateDelaunayTriangles(rect, points):

  # Create an instance of Subdiv2D
  subdiv = cv2.Subdiv2D(rect)

  # Insert points into subdiv
  for p in points:
    subdiv.insert((int(p[0]), int(p[1])))

  # Get Delaunay triangulation
  triangleList = subdiv.getTriangleList()

  # Find the indices of triangles in the points array
  delaunayTri = []

  for t in triangleList:
    # The triangle returned by getTriangleList is
    # a list of 6 coordinates of the 3 points in
    # x1, y1, x2, y2, x3, y3 format.
    # Store triangle as a list of three points
    pt = []
    pt.append((t[0], t[1]))
    pt.append((t[2], t[3]))
    pt.append((t[4], t[5]))

    pt1 = (t[0], t[1])
    pt2 = (t[2], t[3])
    pt3 = (t[4], t[5])

    if rectContains(rect, pt1) and rectContains(rect, pt2) and rectContains(rect, pt3):
      # Variable to store a triangle as indices from list of points
      ind = []
      # Find the index of each vertex in the points list
      for j in range(0, 3):
        for k in range(0, len(points)):
          if(abs(pt[j][0] - points[k][0]) < 1.0 and abs(pt[j][1] - points[k][1]) < 1.0):
            ind.append(k)
        # Store triangulation as a list of indices
      if len(ind) == 3:
        delaunayTri.append((ind[0], ind[1], ind[2]))

  return delaunayTri

# Apply affine transform calculated using srcTri and dstTri to src and
# output an image of size.
def applyAffineTransform(src, srcTri, dstTri, size):

  # Given a pair of triangles, find the affine transform.
  warpMat = cv2.getAffineTransform(np.float32(srcTri), np.float32(dstTri))

  # Apply the Affine Transform just found to the src image
  dst = cv2.warpAffine(src, warpMat, (size[0], size[1]), None,
             flags=cv2.INTER_LINEAR, borderMode=cv2.BORDER_REFLECT_101)

  return dst

# Warps and alpha blends triangular regions from img1 and img2 to img
def warpTriangle(img1, img2, t1, t2):
  # Find bounding rectangle for each triangle
  r1 = cv2.boundingRect(np.float32([t1]))
  r2 = cv2.boundingRect(np.float32([t2]))

  # Offset points by left top corner of the respective rectangles
  t1Rect = []
  t2Rect = []
  t2RectInt = []

  for i in range(0, 3):
    t1Rect.append(((t1[i][0] - r1[0]), (t1[i][1] - r1[1])))
    t2Rect.append(((t2[i][0] - r2[0]), (t2[i][1] - r2[1])))
    t2RectInt.append(((t2[i][0] - r2[0]), (t2[i][1] - r2[1])))

  # Get mask by filling triangle
  mask = np.zeros((r2[3], r2[2], 3), dtype=np.float32)
  cv2.fillConvexPoly(mask, np.int32(t2RectInt), (1.0, 1.0, 1.0), 16, 0)

  # Apply warpImage to small rectangular patches
  img1Rect = img1[r1[1]:r1[1] + r1[3], r1[0]:r1[0] + r1[2]]

  size = (r2[2], r2[3])

  img2Rect = applyAffineTransform(img1Rect, t1Rect, t2Rect, size)

  img2Rect = img2Rect * mask

  # Copy triangular region of the rectangular patch to the output image
  img2[r2[1]:r2[1]+r2[3], r2[0]:r2[0]+r2[2]] = img2[r2[1]:r2[1]+r2[3], r2[0]:r2[0]+r2[2]] * ((1.0, 1.0, 1.0) - mask)
  img2[r2[1]:r2[1]+r2[3], r2[0]:r2[0]+r2[2]] = img2[r2[1]:r2[1]+r2[3], r2[0]:r2[0]+r2[2]] + img2Rect

apply_filter.py
import mediapipe as mp
import cv2
import math
import numpy as np
import faceBlendCommon as fbc
import csv
import argparse

VISUALIZE_FACE_POINTS = False

filters_config = {
    'anonymous':
        [{'path': "filters/anonymous.png",
          'anno_path': "filters/anonymous_annotations.csv",
          'morph': True, 'animated': False, 'has_alpha': True}],
    'anime':
        [{'path': "filters/anime.png",
          'anno_path': "filters/anime_annotations.csv",
          'morph': True, 'animated': False, 'has_alpha': True}],
    'dog':
        [{'path': "filters/dog-ears.png",
          'anno_path': "filters/dog-ears_annotations.csv",
          'morph': False, 'animated': False, 'has_alpha': True},
         {'path': "filters/dog-nose.png",
          'anno_path': "filters/dog-nose_annotations.csv",
          'morph': False, 'animated': False, 'has_alpha': True}],
    'cat':
        [{'path': "filters/cat-ears.png",
          'anno_path': "filters/cat-ears_annotations.csv",
          'morph': False, 'animated': False, 'has_alpha': True}],
    'jason-joker':
        [{'path': "filters/jason-joker.png",
          'anno_path': "filters/jason-joker_annotations.csv",
          'morph': True, 'animated': False, 'has_alpha': True}],
    'gold-crown':
        [{'path': "filters/gold-crown.png",
          'anno_path': "filters/gold-crown_annotations.csv",
          'morph': False, 'animated': False, 'has_alpha': True}],
    'flower-crown':
        [{'path': "filters/flower-crown.png",
          'anno_path': "filters/flower-crown_annotations.csv",
          'morph': True, 'animated': False, 'has_alpha': True}],
    'glasses':
        [{'path': "filters/glass.png",
          'anno_path': "filters/glasses_annotations.csv",
          'morph': False, 'animated': False, 'has_alpha': True}],  
    'mustache':
        [{'path': "filters/mustache2.png",
          'anno_path': "filters/mustache_annotations.csv",
          'morph': True, 'animated': False, 'has_alpha': True}],          
}


def parse(args=None):
    parser = argparse.ArgumentParser()
    parser.add_argument('--image', dest='image', type=str, default='custom/5.jpg')
    parser.add_argument('--attr', dest='attrnum', type=int, default=1)
    return parser.parse_args(args)




# detect facial landmarks in image
def getLandmarks(img):
    mp_face_mesh = mp.solutions.face_mesh
    selected_keypoint_indices = [127, 93, 58, 136, 150, 149, 176, 148, 152, 377, 400, 378, 379, 365, 288, 323, 356, 70, 63, 105, 66, 55,
                 285, 296, 334, 293, 300, 168, 6, 195, 4, 64, 60, 94, 290, 439, 33, 160, 158, 173, 153, 144, 398, 385,
                 387, 466, 373, 380, 61, 40, 39, 0, 269, 270, 291, 321, 405, 17, 181, 91, 78, 81, 13, 311, 306, 402, 14,
                 178, 162, 54, 67, 10, 297, 284, 389]

    height, width = img.shape[:-1]

    with mp_face_mesh.FaceMesh(max_num_faces=1, static_image_mode=True, min_detection_confidence=0.5) as face_mesh:

        results = face_mesh.process(cv2.cvtColor(img, cv2.COLOR_BGR2RGB))

        if not results.multi_face_landmarks:
            print('Face not detected!!!')
            return 0

        for face_landmarks in results.multi_face_landmarks:
            values = np.array(face_landmarks.landmark)
            face_keypnts = np.zeros((len(values), 2))

            for idx,value in enumerate(values):
                face_keypnts[idx][0] = value.x
                face_keypnts[idx][1] = value.y

            # Convert normalized points to image coordinates
            face_keypnts = face_keypnts * (width, height)
            face_keypnts = face_keypnts.astype('int')

            relevant_keypnts = []

            for i in selected_keypoint_indices:
                relevant_keypnts.append(face_keypnts[i])
            return relevant_keypnts
    return 0


def load_filter_img(img_path, has_alpha):
    # Read the image
    img = cv2.imread(img_path, cv2.IMREAD_UNCHANGED)

    alpha = None
    if has_alpha:
        b, g, r, alpha = cv2.split(img)
        img = cv2.merge((b, g, r))

    return img, alpha


def load_landmarks(annotation_file):
    with open(annotation_file) as csv_file:
        csv_reader = csv.reader(csv_file, delimiter=",")
        points = {}
        for i, row in enumerate(csv_reader):
            # skip head or empty line if it's there
            try:
                x, y = int(row[1]), int(row[2])
                points[row[0]] = (x, y)
            except ValueError:
                continue
        return points


def find_convex_hull(points):
    hull = []
    hullIndex = cv2.convexHull(np.array(list(points.values())), clockwise=False, returnPoints=False)
    addPoints = [
        [48], [49], [50], [51], [52], [53], [54], [55], [56], [57], [58], [59],  # Outer lips
        [60], [61], [62], [63], [64], [65], [66], [67],  # Inner lips
        [27], [28], [29], [30], [31], [32], [33], [34], [35],  # Nose
        [36], [37], [38], [39], [40], [41], [42], [43], [44], [45], [46], [47],  # Eyes
        [17], [18], [19], [20], [21], [22], [23], [24], [25], [26]  # Eyebrows
    ]
    hullIndex = np.concatenate((hullIndex, addPoints))
    for i in range(0, len(hullIndex)):
        hull.append(points[str(hullIndex[i][0])])

    return hull, hullIndex


def load_filter(filter_name):

    filters = filters_config[filter_name]

    multi_filter_runtime = []

    for filter in filters:
        temp_dict = {}

        img1, img1_alpha = load_filter_img(filter['path'], filter['has_alpha'])

        temp_dict['img'] = img1
        temp_dict['img_a'] = img1_alpha

        points = load_landmarks(filter['anno_path'])

        temp_dict['points'] = points

        if filter['morph']:
            # Find convex hull for delaunay triangulation using the landmark points
            hull, hullIndex = find_convex_hull(points)

            # Find Delaunay triangulation for convex hull points
            sizeImg1 = img1.shape
            rect = (0, 0, sizeImg1[1], sizeImg1[0])
            dt = fbc.calculateDelaunayTriangles(rect, hull)

            temp_dict['hull'] = hull
            temp_dict['hullIndex'] = hullIndex
            temp_dict['dt'] = dt

            if len(dt) == 0:
                continue

        if filter['animated']:
            filter_cap = cv2.VideoCapture(filter['path'])
            temp_dict['cap'] = filter_cap

        multi_filter_runtime.append(temp_dict)

    return filters, multi_filter_runtime


# process input from webcam or video file


args_ = parse()


# Some variables
count = 0
isFirstFrame = True
sigma = 50

iter_filter_keys = iter(filters_config.keys())
# filters, multi_filter_runtime = load_filter(next(iter_filter_keys))

filter_name = "glasses"
if args_.attrnum == 1:
    filter_name="glasses"
if args_.attrnum == 2:
    filter_name="mustache"

print("filter", filter_name)
filters, multi_filter_runtime = load_filter(filter_name)
# The main loop

frame = cv2.imread(args_.image)

points2 = getLandmarks(cv2.cvtColor(frame, cv2.COLOR_BGR2RGB))


################ Optical Flow and Stabilization Code #####################
img2Gray = cv2.cvtColor(frame, cv2.COLOR_BGR2GRAY)

if isFirstFrame:
    points2Prev = np.array(points2, np.float32)
    img2GrayPrev = np.copy(img2Gray)
    isFirstFrame = False

lk_params = dict(winSize=(101, 101), maxLevel=15,
                    criteria=(cv2.TERM_CRITERIA_EPS | cv2.TERM_CRITERIA_COUNT, 20, 0.001))
points2Next, st, err = cv2.calcOpticalFlowPyrLK(img2GrayPrev, img2Gray, points2Prev,
                                                np.array(points2, np.float32),
                                                **lk_params)

# Final landmark points are a weighted average of detected landmarks and tracked landmarks

for k in range(0, len(points2)):
    d = cv2.norm(np.array(points2[k]) - points2Next[k])
    alpha = math.exp(-d * d / sigma)
    points2[k] = (1 - alpha) * np.array(points2[k]) + alpha * points2Next[k]
    points2[k] = fbc.constrainPoint(points2[k], frame.shape[1], frame.shape[0])
    points2[k] = (int(points2[k][0]), int(points2[k][1]))

# Update variables for next pass
points2Prev = np.array(points2, np.float32)
img2GrayPrev = img2Gray
################ End of Optical Flow and Stabilization Code ###############

if VISUALIZE_FACE_POINTS:
    for idx, point in enumerate(points2):
        cv2.circle(frame, point, 2, (255, 0, 0), -1)
        cv2.putText(frame, str(idx), point, cv2.FONT_HERSHEY_SIMPLEX, .3, (255, 255, 255), 1)
    cv2.imshow("landmarks", frame)

for idx, filter in enumerate(filters):

    filter_runtime = multi_filter_runtime[idx]
    img1 = filter_runtime['img']
    points1 = filter_runtime['points']
    img1_alpha = filter_runtime['img_a']

    if filter['morph']:

        hullIndex = filter_runtime['hullIndex']
        dt = filter_runtime['dt']
        hull1 = filter_runtime['hull']

        # create copy of frame
        warped_img = np.copy(frame)

        # Find convex hull
        hull2 = []
        for i in range(0, len(hullIndex)):
            hull2.append(points2[hullIndex[i][0]])

        mask1 = np.zeros((warped_img.shape[0], warped_img.shape[1]), dtype=np.float32)
        mask1 = cv2.merge((mask1, mask1, mask1))
        img1_alpha_mask = cv2.merge((img1_alpha, img1_alpha, img1_alpha))

        # Warp the triangles
        for i in range(0, len(dt)):
            t1 = []
            t2 = []

            for j in range(0, 3):
                t1.append(hull1[dt[i][j]])
                t2.append(hull2[dt[i][j]])

            fbc.warpTriangle(img1, warped_img, t1, t2)
            fbc.warpTriangle(img1_alpha_mask, mask1, t1, t2)

        # Blur the mask before blending
        mask1 = cv2.GaussianBlur(mask1, (3, 3), 10)

        mask2 = (255.0, 255.0, 255.0) - mask1

        # Perform alpha blending of the two images
        temp1 = np.multiply(warped_img, (mask1 * (1.0 / 255)))
        temp2 = np.multiply(frame, (mask2 * (1.0 / 255)))
        output = temp1 + temp2
    else:
        dst_points = [points2[int(list(points1.keys())[0])], points2[int(list(points1.keys())[1])]]
        tform = fbc.similarityTransform(list(points1.values()), dst_points)
        # Apply similarity transform to input image
        trans_img = cv2.warpAffine(img1, tform, (frame.shape[1], frame.shape[0]))
        trans_alpha = cv2.warpAffine(img1_alpha, tform, (frame.shape[1], frame.shape[0]))
        mask1 = cv2.merge((trans_alpha, trans_alpha, trans_alpha))

        # Blur the mask before blending
        mask1 = cv2.GaussianBlur(mask1, (3, 3), 10)

        mask2 = (255.0, 255.0, 255.0) - mask1

        # Perform alpha blending of the two images
        temp1 = np.multiply(trans_img, (mask1 * (1.0 / 255)))
        temp2 = np.multiply(frame, (mask2 * (1.0 / 255)))
        output = temp1 + temp2

    frame = output = np.uint8(output)

cv2.putText(frame, "Press F to change filters", (10, 20), cv2.FONT_HERSHEY_SIMPLEX, .5, (255, 0, 0), 1)
cv2.imshow("Face Filter", output)
cv2.imwrite("result.jpg", output)
cv2.waitKey(0)

# keypressed = cv2.waitKey(1) & 0xFF
# if keypressed == 27:
#     break
# elif keypressed == ord('f'):
#     try:
#         filters, multi_filter_runtime = load_filter(next(iter_filter_keys))
#     except:
#         iter_filter_keys = iter(filters_config.keys())
        # filters, multi_filter_runtime = load_filter(next(iter_filter_keys))

count += 1

cv2.destroyAllWindows()